[Team03][BE] 중간점검 PR 드립니다!

pbg0205 opened the pull request:
안녕하세요. Team03의 K.🧑‍💻, Cooper.🧑‍💻입니다.
중간점검 및 리뷰받고자 PR 보내드리게 되었습니다.
구현한 기능은 아래와 같습니다.

## 😀 기능 구현을 하면서 중점을 둔 컨셉
각각의 도메인들이 **별도의 aggregate**로 구성하였습니다. 그 이유는 각각의 도메인마다 **생성주기가 모두 다르다**고 생각했습니다.
그래서 별도의 도메인으로 작업을 진행하고 issue에서 해당 기능을 모두 통합하는 방식으로 진행하였습니다.
현재 조건에 따른 `Issue filter`, `이미지 업로드 기능`를 제외한 요구사항에 해당하는 기능을 구현한 상태입니다.
<br>

### 🔐 github 소셜 로그인
- github server에서 access token을 발급과 함께 유저 정보를 가져와 JWT로 변환하는 로직을 구현하였습니다.
- login 시, 유저 정보를 DB에서 존재 유무를 판별하여, 존재하지 않을 경우에 유저 정보를 추가하도록 하였습니다.
- 모든 기능을 이용 시, interceptor를 통해 authorization과정을 통해 사용자를 확인하여 기능을 사용할 수 있도록 하였습니다.
<br>

### 🤖  마일스톤 기본 기능
- 마일스톤 전체 조회
- 마일스톤 추가 기능
- 마일스톤 수정 기능
- 마일스톤 삭제 기능
<br>

### 🎃 라벨 기본 기능
- 라벨 전체 조회 기능
- 라벨 추가 기능
- 라벨 수정 기능
- 라벨 삭제 기능
<br>

### 😺 이슈 기본 기능
- 이슈 추가 기능
   - 이슈 생성 시, 담당자, 라벨, 마일스톤을 등록 가능
- 이슈 수정 기능
   - 이슈 제목 수정 가능
   - 이슈 상태(open/closed) 변경 기능
- 담당자, 라벨, 마일스톤 기능
<br>

## 🤔 고민했던 점 + 접근(해결) 방법

1. 현재, 이슈와 담당자, 이슈와 라벨의 관계 테이블을 엔티티로 두지 않고 `@ManyToMany` 관계로 사용했습니다.
2. 별도의 aggregate로 접근했을 때는 `@OneToMany`와 `Lazy Loading`을 사용하는 것보다 별도 조회 방식이 적합하다고 생각했습니다.
3. label 제거 시, 참조 관계로 인해 발생하는 삭제 불가 문제는 mysql migration 과정에 DDL 작성 과정에서 해결할 예정입니다.
<br>
위와 같이 정리하여 중간점검 PR 드립니다.<br>
바쁘신 와중에 시간내어 리뷰해주셔서 감사합니다!<br>
좋은하루 보내세요. 😀

pbg0205 committed:
feat : BE first commit

pbg0205 committed:
feat : Add Jwt Resonse.(#2)

kihyuk-sung committed:
feat: Issue Jwt for github web login user (#2)

- github 로그인을 하면, jwt 발행하여 보내줍니다.

kihyuk-sung committed:
feat: CORS Configuration

- CORS 설정
- localhost:3000에 여러 method 허용

pbg0205 committed:
feat : GitHub ios for login(#2)

kihyuk-sung committed:
Merge pull request #18 from pbg0205/BE/feat/github-login

[Be] github login 기능 구현

kihyuk-sung committed:
feat: Create new label (#13)

- 새 라벨을 생성하고, 추가되는지 확인.
- Validation check는 아직 없음.
- Color에 대해서 어떻게 저장할 지 협의 해야합니다.

kihyuk-sung committed:
feat: Colors value object (#13)

- 색상 정보를 Colors로 옮겼습니다.
- label의 name으로도 찾을 수 있고, 색상으로 묶어서 관리하면 객체 유효성
판단를 쉽게 할 수 있다고 생각했습니다.

kihyuk-sung committed:
refactor: Label (#13)

- 색상을 표시하는 값 객체로 Colors 객체를 사용
- Label.create에 대해서 테스트

kihyuk-sung committed:
test: Add LabelTest (#13)

- Label을 생성할 수 있다.
- Label의 이름이 null이거나 empty이면 생성할 수 없다.

kihyuk-sung committed:
test: Label name & description length check (#13)

- name 50자 이하, description 100자 이하

kihyuk-sung committed:
feat: Label name & description check (#13)

- 이제 name과 description 길이 체크합니다.
- 각각 50자, 100자

kihyuk-sung committed:
feat: Label update

- 라벨 업데이트

kihyuk-sung committed:
feat: LabelRepository (#13)

kihyuk-sung committed:
feat: LabelService (#13)

kihyuk-sung committed:
feat: LabelController (#13)

- GET /api/labels
- POST /api/labels
- PUT /api/labels/:id
- DELETE /api/labels/:id

kihyuk-sung committed:
fix: Add transactional for dirty check (#13)

- dirty check해서 자동으로 저장하도록 애노테이션 추가

kihyuk-sung committed:
feat: Add table constraint (#13)

- name 길이, description 길이, name unique 제약

kihyuk-sung committed:
refactor: Move wrapper into service (#13)

- wrapping을 service에서 하도록 함.

kihyuk-sung committed:
style: code formatting (#13)

pbg0205 committed:
chore : Add test dependency and h2 properties(#14)

pbg0205 committed:
feat : Add Milestone domain(#14)

pbg0205 committed:
feat : Add MilestoneRepository(#14)

pbg0205 committed:
feat : Add H2 DataBase connection(#14)

pbg0205 committed:
feat : Add dto and wrapper(#14)

pbg0205 committed:
feat : Add milestone CRUD(#14)

kihyuk-sung committed:
feat: review 반영 (#13)

- 될 수 있으면 롬복으로 객체 생성, 설정하도록 함
- 네이밍 컨벤션 맞춤

pbg0205 committed:
feat : Correct detail of field option (#14)

- Add nullable false in title

- Add nullable false in description

- Remove @NonNull in dueDate

- Add access modifier in NoArgsConstructor annotation

pbg0205 committed:
feat : Correct details in MilestoneRequestDt(#14)

- Add final keyword in field

- Correct constructor to RequiredArgsConstructor annotation

pbg0205 committed:
style : Correct variable name(#14)

- Correct MilestoneRequest to MilestoneRequestDto

pbg0205 committed:
refactor : Remove Junit implementation(#14)

pbg0205 committed:
refactor : Correct details in MilestoneResponse(#14)

- Add final keyword in field

- Correct AllArgsConstructor access modifier

- Correct AllArgsConstructor to AllargsConstructor annotation

pbg0205 committed:
refactor : Correct DB Column name(#14)

pbg0205 committed:
refactor : Add AllArgsConstructor annotation(#14)

pbg0205 committed:
refactor : Correct logic to method reference(#14)

pbg0205 committed:
refactor : Correct method name(#14)

kihyuk-sung committed:
Merge pull request #33 from pbg0205/BE/feat/label

[BE] Label CRUD

kihyuk-sung committed:
Merge branch 'BE-dev' into BE/feat/milestone

kihyuk-sung committed:
Merge pull request #38 from pbg0205/BE/feat/milestone

[BE] Milestone CRUD

pbg0205 committed:
refactor : Correct user to auth package(#19))

pbg0205 committed:
feat : Add user domain(#19)

pbg0205 committed:
refactor : Change GitHubUser(#19)

pbg0205 committed:
feat : Add static factory method(#19)

pbg0205 committed:
feat : Add UserRepository(#19)

pbg0205 committed:
feat : Add UserService(#19)

kihyuk-sung committed:
feat: Add UserController (#19)

- UserController 추가
- UsersWrapper 추가
- UserService에서 UserDto를 wrap

kihyuk-sung committed:
feat: Jwt User claim (#19)

- 이제 JWT에 완전한 user claim을 담습니다.
- id, nickname, imageUrl

pbg0205 committed:
Merge pull request #41 from pbg0205/BE/feat/user

[BE] User Domain

kihyuk-sung committed:
refactor: Change Milestone Wrapper field name

- milestone과 milestones로 수정

kihyuk-sung committed:
refactor: Change label.infra packages

- 패키지 이름 수정

pbg0205 committed:
feat : Add Comment domain(#45)

pbg0205 committed:
feat : Add CommentRepository(#45)

pbg0205 committed:
feat : Add Comment domain and dto

pbg0205 committed:
feat : Add CommentWrapper and CommentsWraper

pbg0205 committed:
feat : Add Comment's create, update(#45)

pbg0205 committed:
feat : Add CommentController(#45)

- Comment 추가 기능 구현

- Comment 수정 기능 구현

kihyuk-sung committed:
Merge pull request #46 from pbg0205/BE/feat/comment

[BE] comment 기능 구현

kihyuk-sung committed:
fix: Comment.createdAt (#10)

- `@CreatedDate` 동작 안되서, 삭제하고 생성 시간 추가하는 로직 구현

kihyuk-sung committed:
feat: User.equals and hashcode (#10)

- id로 equals, hashcode 비교하도록 구현

kihyuk-sung committed:
feat: Issue 기본 생성 (#10)

- 코멘트 없이, Issue 생성하여 DB 저장 후, 가져오는 것까지 확인

kihyuk-sung committed:
refactor: Add JoinColumn to comment (#10)

- OneToMany만 있으면 중간 테이블 생성됩니다.
- Issue - Comment는 OneToMany로, Comment에 IssueId를 갖도록 함.

kihyuk-sung committed:
refactor: Remove Comment List in Issue

- 이슈 내에서 OneToMany Comment 삭제

kihyuk-sung committed:
refactor: Add comment in Comment Object

- issueId에 외래키 제약 조건에 대한 주석 추가

kihyuk-sung committed:
feat: Add Issue Dto

- Issue를 위한 Dto를 추가하였습니다.

kihyuk-sung committed:
feat: IssueService

- IssueService.createIssue 추가

kihyuk-sung committed:
feat: IssueController

- IssueController 추가
- POST /api/issues

kihyuk-sung committed:
feat: AuthInterceptor

- Authorization Header를 확인하여, User를 추가하는 Interceptor
- User 생성 메소드 추가
- JwtUtils에 토큰 파싱하는 로직 추가

kihyuk-sung committed:
fix: User bug if nickname is null

- GitHub 로그인 시에 nickname이 null일 수 있다.
- `@NonNull` 삭제

kihyuk-sung committed:
feat: Add Interceptor

- Interceptor를 추가하였습니다.
- CORS Preflight를 위해서 Options는 Interceptor에서 허용

kihyuk-sung committed:
fix: Allow /api/login if Authorization is null

- /api/login은 Authorization 헤더 없어도 허용

kihyuk-sung committed:
fix: Create Issue

- Issue를 생성할 수 있습니다.

pbg0205 committed:
feat : Correct comment update function(#45)

pbg0205 committed:
refacotr : Correct name repository to infra(#10)

pbg0205 committed:
feat : Update and Read issue(#10)

- 이슈 제목을 수정할 수 있습니다.

- 이슈를 이슈 번호를 통해 조회할 수 있습니다.

pbg0205 committed:
feat : Change open/closed status(#10)

pbg0205 committed:
feat : Add and remove label in issue(#10)

kihyuk-sung committed:
feat : Add and remove assignee in issue(#10)

kihyuk-sung committed:
feat: Add Milestone to Issue (#10)

kihyuk-sung committed:
feat: Remove milestone from Issue (#10)

kihyuk-sung committed:
feat: Read All Issues (#10)

- 모든 이슈를 조회하는 기능 추가

pbg0205 committed:
feat : Add Issues Count in Mileston(#10)

pbg0205 committed:
feat : Count open and closed issues(#10)

pbg0205 committed:
feat : Change issueCount and open at update(#10)

pbg0205 committed:
refactor: Change if statement(#10)

kihyuk-sung committed:
Merge pull request #60 from pbg0205/BE/feat/issue

[BE] issue 기본 기능 구현

kihyuk-sung committed:
feat: Migration to MySql

kihyuk-sung committed:
fix: Remove foreign key constraint

- 중간 연결 테이블의 issue에 대해서는 제약이 걸리지 않아 제거

pbg0205 committed:
Merge pull request #65 from pbg0205/BE/feat/mysql

[BE] MySQL DB

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/auth/component/JwtUtils.java:
```diff
@@ -0,0 +1,59 @@
+
+    private static final String ISSUER = "issue-tracker";
+    private static final String ID = "id";
+    private static final String NICKNAME = "nickName";
```
정확히는 nickname 이 맞다고 합니다~

pbg0205 replied:
아 그러네요... 감사합니다! 수정하도록 하겠습니다!

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/auth/component/JwtUtils.java:
```diff
@@ -0,0 +1,59 @@
+    private final Algorithm ALGORITHM;
+    private final JWTVerifier jwtVerifier;
+
+    public JwtUtils(@Value("${auth.jwt.secret}") String SECRET) {
```
value 에 대해선 대문자로 쓰지 않는게 일반적인 것 같습니다

kihyuk-sung replied:
그렇군요. 수정했습니다.

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/auth/component/JwtUtils.java:
```diff
@@ -0,0 +1,59 @@
+    public JwtUtils(@Value("${auth.jwt.secret}") String SECRET) {
+        ALGORITHM = Algorithm.HMAC256(SECRET);
+        jwtVerifier = JWT.require(ALGORITHM)
+                .acceptExpiresAt(0)
```
0 은 어떤의미인가요?

+ required().acceptExpiresAt() 에 관해서는 아래 의미가 있다고 하네요. 0이면 strict 하게 검사하게 될 것 같습니다~ 수정의견은 아니니 잘 판단하셔서 설정하면 좋겠네요!

JWT.require(xxx).acceptExpiresAt(5 * 60) means you will accept a token which has already expired 5 minutes before.Even considering the network lag, 5 minutes of leeway is still too long. It should in seconds.

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/auth/interceptor/AuthInterceptor.java:
```diff
@@ -0,0 +1,42 @@
+        }
+
+        String token = getJwt(request);
+        request.setAttribute("author", jwtUtils.getUserFromJwt(token));
```
author 보다는 authUser가 낫지 않을까 합니다.ㅎㅎ

pbg0205 replied:
수정 완료했습니다! 감사합니다!

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/auth/component/JwtUtils.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.UUID;
+
+@Component
+public class JwtUtils {
```
빈 이름으로 Utils는 잘 사용하지 않는데요, 로직에는 상관이 없지만 더 좋은 이름이 있다면 고민해보시는 것을 권해드립니다~

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/auth/interceptor/AuthInterceptor.java:
```diff
@@ -0,0 +1,42 @@
+        if (!authorizationHeader.startsWith("Bearer")) {
+            throw new RuntimeException("토큰 타입 이상");
+        }
+        return authorizationHeader.substring("Bearer".length()).trim();
```
string의 replace 메서드를 활용할 수도 있겠습니다.

kihyuk-sung replied:
replace가 좀더 의미가 명확하네요! 그래도 혹시나 토큰의 내용에 Bearer가 될 수도 있으니 replaceFirst를 이용하였습니다.

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/auth/service/GithubLoginService.java:
```diff
@@ -0,0 +1,99 @@
+        this.userRepository = userRepository;
+    }
+
+    public JwtResponse issueToken(String code, GitHubOauthValues gitHubOauthValues) {
```
private 이어도 되지 않을까요?

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/comment/domain/Comment.java:
```diff
@@ -0,0 +1,57 @@
+        return new Comment(issueId, author, content);
+    }
+
+    public void updateContent(CommentRequest commentRequest) {
+        this.content = commentRequest.getContent();
+    }
```
CommentRequest 클래스가 Comment 도메인까지 엮일 필요는 없다고 생각이 듭니다~ 상위에서 content만 빼서 인자로 넘겨주는 것은 어떨까요?

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/comment/dto/CommentResponse.java:
```diff
@@ -0,0 +1,33 @@
+
+    private final String content;
+
+    public static CommentResponse fromEntity (Comment comment) {
+        return new CommentResponse(comment.getId(),
+                comment.getIssueId(),
+                UserDto.fromEntity(comment.getAuthor()),
+                comment.getCreatedAt(),
+                comment.getContent()
+        );
+    }
```
static 메서드로 객체 생성 로직을 담는 것은 권장되는 패턴이지만, 요 메서드가 호출되는 시점에 comment.getAuthor() 가 제대로 동작한다고 보장할 수 있을까요? comment 엔티티의 author엔티티가 트랜잭션이 끝나기 전에 로드가 되었다고 보장하기 힘들 것 같은데요.. 물론 잘못되었다면 테스트 중에 수정이 되겠지만 요런 측면도 있다~ 까지만 챙겨주세요ㅎㅎ

pbg0205 replied:
넵 알겠습니다! 감사합니다!

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/comment/service/CommentService.java:
```diff
@@ -0,0 +1,40 @@
+        this.commentRepository = commentRepository;
+    }
+
+    public CommentWrapper createComment(CommentRequest commentRequest, User author) {
+        Comment comment = commentRequest.toEntity(author);
+        Comment savedComment = commentRepository.save(comment);
+        return CommentWrapper.wrap(CommentResponse.fromEntity(savedComment));
+    }
```
여기 Transactional 어노테이션이 붙어야할 이유는 없을까요?

kihyuk-sung replied:
누락됐네요! 수정했습니다.

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/issue/domain/Issue.java:
```diff
@@ -0,0 +1,137 @@
+    public void setMilestone(Milestone milestone) {
+        this.milestone = milestone;
+
+        if(this.open) {
+            this.milestone.countOpenIssuesUp();
+        }
+
+        if(!this.open) {
+            this.milestone.countClosedIssuesUp();
+        }
```
저라면 If else 로 하겠습니다~

LarryJung commented on BE/src/main/java/com/codesquad/issuetracker/milestone/domain/Milestone.java:
```diff
@@ -0,0 +1,82 @@
+        this.dueDate = milestoneRequest.getDueDate();
+    }
+
+    public void countOpenIssuesUp() {
+        this.openIssues++;
+    }
+
+    public void countClosedIssuesUp() {
+        this.closedIssues++;
+    }
+
+    public void countOpenIssuesDown() {
+        this.openIssues--;
+    }
+
+    public void countClosedIssuesDown() {
+        this.closedIssues--;
+    }
+
+    public void openIssue() {
+        this.openIssues++;
+        this.closedIssues--;
+    }
+
+    public void closeIssue() {
+        this.openIssues--;
+        this.closedIssues++;
+    }
```
이 숫자 연산이 한명의 유저에 종속된건 아니고 다수의 유저로부터 동시에 요청될 수 있을텐데요, db 락 가능성에 대해 생각해보시면 좋을 것 같습니다~

LarryJung commented on BE/src/test/java/com/codesquad/issuetracker/issue/infra/IssueRepositoryTest.java:
```diff
@@ -0,0 +1,102 @@
+class IssueRepositoryTest {
+
+    @Autowired
+    IssueRepository issueRepository;
```
private 으로 둬도 될듯합니다!

pbg0205 replied:
넵 수정 완료하였습니다! 감사합니다!

LarryJung reviewed (approved):
안녕하세요~ 기능 분리도 잘 되어있고 이모저모 좋은 습관으로 짜고 계신게 느껴져서 좋았습니다. 👍👍

하지만 제가 몇개 코멘트 남겨드린 부분은 중요하므로 잘 생각해주시면 좋겠구요, 대표로 먼저 보였던 위치에서 남긴 것이기 때문에 다른 부분에도 동일하게 피드백 반영이 되면 좋겠습니다.

추가로 예외 처리는 조금 미흡한 것 같아서 말씀드립니다~ global exception handler 를 만들어서 예외상황에서 일관된 방식으로 응답을 주도록 수정이 되면 클라이언트와 예외처리 방식에 대해서 협의하기도 편하고 각자 다른 방식으로 예외처리를 하지 않게 되니 장점이 있으니 적용해 보세요!

pbg0205 committed:
refactor : Correct nickName to nickname

kihyuk-sung committed:
refactor: Change variables uppercase to camelcase

- `@Value`로 받는 변수의 경우는 대문자를 보통 사용하지 않는다고 합니다

pbg0205 committed:
refactor : Correct author to authUser

kihyuk-sung committed:
refactor: Change substring() to replaceFirst()

- replace가 좀더 의미가 명확해서, 이 메소드를 이용
- 이후에 토큰이 Bearer라고 달릴 수도 있으므로, replaceFirst로 사용

kihyuk-sung committed:
refactor: Add `@Transactional` to createComment

pbg0205 committed:
refactor : Correct if to else if

pbg0205 committed:
refactor : Add private in MilestRepository

pbg0205 commented:
리뷰해주셔서 감사합니다! @LarryJung
리뷰해주신 부분 locking과 JwtUtils부분 외에 수정 완료하였습니다.
리뷰 반영 중에 locking부분에서 `낙관적 잠금`과 `비관적 잠금`을 학습하였습니다.
 저희의 경우, update하는 과정에서 일관성이 유지되기 위해서는 다른 사용자가 update를 할 수 없는 방식이 적합하다고 생각했습니다.
그래서 비관적 잠금을 도입하고자 하였는데 연관된 domain까지 locking을 걸어주는 과정에서 적절한 해결방안을 찾지 못하여 일단 보류해두었습니다ㅠ 아직 메인 기능 구현이 부족하여 추후에 JwtUtils 부분 또한 수정하도록 하겠습니다!! 
바쁘시겠지만 검토 부탁드리겠습니다.
감사합니다 😀

LarryJung reviewed (approved):
추가 수정분까지 확인했습니다. 일단 승인 드립니다.^^  메인 기능 구현해주세요~!
